### DISCUSSION TITLE
Why is the first target forced to be an absolute joint target?

### DISCUSSION THREAD

**Discussion by kevingessner** [visose/Robots#272] [Q&A] [answered]
During program processing, the first target is changed from a cartesian target to a joint target (i.e. from a MoveL/MoveJ in rapid to a MoveAbsJ):

https://github.com/visose/Robots/blob/ac190db1eb17fb0639d289e28deb8ab1f90ebb41/src/Robots/Program/CheckProgram.cs#L47-L48

This was added ages ago in commit e4f294454cc8ba8c8b305851ac17118c3ab4d8ab, with no comment or indication of why it's necessary.

Is it (still) necessary? It shouldn't be required by the robot or by the kinematics solver, as far as I can tell. It can cause a problem because the first target thus doesn't respect the frame/work object, and is hard to debug and change later in the code. If it's not required, could that "fix" be removed or made optional?

**Comment by visose** [comment, marked as answer]
It was added because the position of the real robot is unknown in Grasshopper and depending on where it is, could adopt a different pose when reaching the first target, which would not match the simulation. A joint target would always adopt same pose no matter where the robot was before.
With some robots, like ABB, you could use a `confdata` variable and a cartesian joint target to unambiguously define it, but not all robots have this.
The usual practice is to define a "home position" as the first target that is safe place and wouldn't need to be defined relative to a workobject.
